docs/proposal-0007-discrete_metric_execution.md

```diff
@@ -0,0 +1,107 @@
+* **-** More total queries (by factor of `m` where `m` is the # of metrics (or # of metric data sources, possibly))
+
+#### Query Analysis
+
+I picked a recent experiment, and ran a single metrics query with 4 data sources, and then 4 discrete metrics queries, one for each data source (full query and data [here](https://docs.google.com/document/d/1wcHTsnG75oeaABWswcnu_nv4RfKf5647LlaClJFi84c/edit#heading=h.73nlojxheztt)). Here are the takeaways:
+
+For the single query vs sum of four queries:
+* Total bytes processed (per BigQuery console before execution) was the same
+* Bytes shuffled was the same
+* Elapsed time was ~33% greater for the 4 queries
+* Slot time was ~50% less for the 4 queries
+
+
+### Analysis Period Dependencies
+
+The idea of this component is to define dependencies between analysis periods so that they can be computed in a particular order. This would allow us to:
+* Compute multi-day analysis periods based off of single day analysis period results instead of re-computing from scratch
+* Compute pre-enrollment periods before post-enrollment periods (for covariate adjustment)
+
+#### Questions
+* should this be configurable? if so, configured by experiment, app, etc.?
+* caveats to computing multi-day analysis from daily?
+
+#### Pros / Cons
+
+* **+** Guarantee that pre-enrollment analysis periods exist when needed
+* **+** Save re-querying (potentially large) telemetry tables for the same information twice (by computing multi-day periods from daily results)
+* **+** Help to prevent loss of fidelity from rerunning an experiment after clients have been shredded (by computing multi-day periods from daily results)
+
+
+### Cache Intermediate Data
+
+In addition to potentially using the daily results as a cache for larger analysis periods (see [Analysis Period Dependencies](#Analysis-Period-Dependencies) above), and the enrollments table as a cache of enrolled clients (current behavior), we can cache additional information to avoid having to construct a single complex query that joins all the data from various sources. Instead, by caching some more of the data from external sources, we can simplify the joins required for metrics results.
+
+#### Questions
+
+* What should we cache? Some options:
+  * enrollments + exposures by analysis period + window
+    * this is the `enrollments` CTE at the beginning of the metrics query
+  * metrics data sources (filtered to experiment dates + existing filters)
+    * this seems potentially huge
+    * data sources are user-defined (so this may be ill-advised)
+  * if we do both of the above could we potentially cluster on analysis window index (or date) to improve join performance?
+
+#### Pros / Cons
+
+* **+** Fewer queries against larger external tables
+* **+** Optimizes for many [discrete metric queries](#discrete-metric-execution)
+* **+** Reduce complexity of metric query joins
+* **-** More total tables / storage needed
```

Comment:
- Also increases the complexity in logic in jetstream

I feel like this part is only worth considering if we keep hitting issues after implementing the things mentioned above

Reply:
Yea this is a good point. I'll add it to the pros/cons list for posterity, and agreed that there's value in doing these things in a staged way so that we're able to evaluate their effectiveness individually, and avoid doing more than necessary.